### PR TITLE
fix(ecs-patterns): allow imported load balancers as inputs

### PR DESCRIPTION
**Pull request by bvtujo**
This PR adds functionality in aws-ecs-patterns to allow imported Network and Application load balancers to be used when initializing the LoadBalancedEC2Service and LoadBalancedFargateService constructs. By necessity, this means:

1) editing the Network and Application load balancer constructs to add an optional IVpc property to be used when calling the from<LBType>LoadBalancerAttributes static methods, and 
2) changing the error which is thrown when calling addTargets on imported load balancers. 

Unit tests are added in aws-ecs-patterns to test the new import functionality and in aws-elasticloadbalancingv2 to ensure that addTargets behaves properly when a Vpc is or is not specified for imported load balancers. 

Resolves: https://github.com/aws/aws-cdk/issues/5209

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Commit Message
fix(ecs-patterns): allow imported load balancers as inputs

This PR adds functionality in aws-ecs-patterns to allow imported Network and Application load balancers to be used when initializing the LoadBalancedEC2Service and LoadBalancedFargateService constructs. By necessity, this means:

1) editing the Network and Application load balancer constructs to add an optional IVpc property to be used when calling the from<LBType>LoadBalancerAttributes static methods, and 
2) changing the error which is thrown when calling addTargets on imported load balancers. 

Unit tests are added in aws-ecs-patterns to test the new import functionality and in aws-elasticloadbalancingv2 to ensure that addTargets behaves properly when a Vpc is or is not specified for imported load balancers. 

Resolves: https://github.com/aws/aws-cdk/issues/5209

### CODE REVIEW

File: packages/@aws-cdk/aws-ecs-patterns/test/ec2/test.l3s.ts [hunk 1]
```diff
@@ -898,4 +897,147 @@ export = {
...
+
+    // WHEN
+    new ecsPatterns.ApplicationLoadBalancedEc2Service(stack, 'Service', {
+      cluster,
+      loadBalancer: alb,
+      taskDefinition: taskDef,
+    });
+
+    // THEN
+    expect(stack).to(haveResourceLike('AWS::ECS::Service', {
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/test/ec2/test.l3s.ts]: You should check the lb resources

File: packages/@aws-cdk/aws-ecs-patterns/test/ec2/test.l3s.ts [hunk 2]
```diff
@@ -898,4 +897,147 @@ export = {
...
+
+    // WHEN
+    new ecsPatterns.NetworkLoadBalancedEc2Service(stack, 'Service', {
+      cluster,
+      loadBalancer: nlb,
+      taskDefinition: taskDef,
+    });
+
+    // THEN
+    expect(stack).to(haveResourceLike('AWS::ECS::Service', {
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/test/ec2/test.l3s.ts]: Can you validate the loadbalancer resources here?

File: packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.load-balanced-fargate-service.ts [hunk 3]
```diff
@@ -412,4 +412,151 @@ export = {
...
+    new ecsPatterns.NetworkLoadBalancedFargateService(stack, "Service", {
+      vpc,
+      loadBalancer: nlb,
+      taskImageOptions: {
+        image: ecs.ContainerImage.fromRegistry('amazon/amazon-ecs-sample'),
+      },
+    });
+
+    // THEN
+    expect(stack).to(haveResourceLike('AWS::ECS::Service', {
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.load-balanced-fargate-service.ts]: Same comment as above.

File: packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.load-balanced-fargate-service.ts [hunk 4]
```diff
@@ -412,4 +412,151 @@ export = {
...
+    new ecsPatterns.ApplicationLoadBalancedFargateService(stack, "Service", {
+      cluster,
+      loadBalancer: alb,
+      taskImageOptions: {
+        image: ecs.ContainerImage.fromRegistry('amazon/amazon-ecs-sample'),
+      }
+    });
+
+    // THEN
+    expect(stack).to(haveResourceLike('AWS::ECS::Service', {
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.load-balanced-fargate-service.ts]: Same comment as above.

**Comment by pkandasamy91** [comment]
@rix0rrr this PR contains changes to the LoadBalancer constructs, please review

**Comment by pkandasamy91** [comment]
@bvtujo did you try deploying this? How does it react when you specify a vpc as part of the LB attributes that is not the vpc that was used to create the load balancer?

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-load-balancer.ts [hunk 5]
```diff
@@ -38,6 +38,14 @@ export interface NetworkLoadBalancerAttributes {
    * @default - When not provided, LB cannot be used as Route53 Alias target.
    */
   readonly loadBalancerDnsName?: string;
+
+  /**
+   * The optional VPC to associate with the imported load balancer.
+   *
+   * @default - When not provided, imported load balancers cannot be used
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-load-balancer.ts]: This message refers to ECS patterns, can we make this more generic?

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/shared/base-load-balancer.ts [hunk 6]
```diff
@@ -109,12 +109,9 @@ export abstract class BaseLoadBalancer extends Resource {
   public readonly loadBalancerSecurityGroups: string[];
 
   /**
-   * The VPC this load balancer has been created in, if available
-   *
-   * If the Load Balancer was imported, the VPC is not available.
+   * the vpc to create the load balancer inside
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-elasticloadbalancingv2/lib/shared/base-load-balancer.ts]: You might want to leave this as `The VPC this load balancer has been created in`. These are the properties of the construct, not the input props.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts [hunk 7]
```diff
@@ -124,7 +125,7 @@ export interface ApplicationLoadBalancedServiceBaseProps {
    *
    * @default - a new load balancer will be created.
    */
-  readonly loadBalancer?: ApplicationLoadBalancer;
+  readonly loadBalancer?: IApplicationLoadBalancer;
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: Can you add a comment here that the VPC attribute of the imported load balancer needs to be specified in order to be used?

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts [hunk 8]
```diff
@@ -102,7 +102,7 @@ export interface NetworkLoadBalancedServiceBaseProps {
    *
    * @default - a new load balancer will be created.
    */
-  readonly loadBalancer?: NetworkLoadBalancer;
+  readonly loadBalancer?: INetworkLoadBalancer;
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts]: Same comment as above.

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-load-balancer.ts [hunk 9]
```diff
@@ -38,6 +38,14 @@ export interface NetworkLoadBalancerAttributes {
    * @default - When not provided, LB cannot be used as Route53 Alias target.
    */
   readonly loadBalancerDnsName?: string;
+
+  /**
+   * The optional VPC to associate with the imported load balancer.
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-load-balancer.ts]: nit: I'd refrain from describing this as optional, and just mention `The VPC to associate to the imported load balancer`

**Comment by rmmeans** [comment]
Has this PR stopped? I'm very interested in being able to use an an imported load balancer with the ECS Patterns. As it stands right now, it looks like I will have to roll my own pattern as creating a new ALB for each app is not cost efficient.

**Comment by bvtujo** [comment]
@rmmeans This PR is still in progress pending internal discussions on how to proceed. 

Do you have an example use case you'd feel comfortable sharing?

**Comment by rmmeans** [comment]
Sure - I can at least give a brief breakdown.

We commonly have a single public ALB that fronts a product, that product might be made up of many microservices, or, we have a single private ALB that fronts all the microservices for that product.

Because this ALB is shared, it is managed outside of the CDK right now. Individual applications (the microservices) each use the CDK to define their respective fargate envs and related auto-scaling of fargate services, roles, alarms, load balancer attachments, etc.

So our use of the CDK will commonly require that when using patterns construct, that it allows us to bring in some of the things on our own (such as the ALB and VPC).

Hope that helps!

**Comment by bvtujo** [comment]
@rmmeans Thanks for the use case details. I'd also be interested in your major deployment concerns and strategies–do you typically need to deploy microservices one at a time, or is it possible to deploy the full breadth of services you require in a single CFN stack? Do you need assurances that deploying a new service won't result in ALB downtime? 

The patterns currently don't support creating multiple services on the same port, as doing so requires that an existing listener and load balancer security group also be imported, but you can hook this up quite straightforwardly with a few lines of code using other CDK modules. 

For an application load balanced Fargate service, the following will do the trick. You will need to import a few things other than the load balancer from the other stack, however, including:
* VPC
* cluster
* load balancer security group
* security groups of any other services running in the cluster which your service needs to communicate with
* listener 
* load balancer

```javascript
// take care of importing all required constructs 
const vpc = ec2.Vpc.fromLookup(this, "vpc", {vpcId: "my-vpc-id"});
const lbSecurityGroup = ec2.SecurityGroup.fromSecurityGroupId(this, "sgId", "my-alb-sg-id");
const importedCluster = ecs.Cluster.fromClusterAttributes(this, "clusterImport", {
  clusterName: "my-cluster",
  vpc,
  securityGroups: [ec2.SecurityGroup.fromSecurityGroupId(this, "sg", "other-service-sg")]; //and other security groups on the cluster
});
const listener = elbv2.ApplicationListener.fromApplicationListenerAttributes(this, "Listener", {
  listenerArn: "my-application-listener-arn",
  securityGroup: lbSecurityGroup,
});
const alb = elbv2.ApplicationLoadBalancer.fromApplicationLoadBalancerAttributes(this, "loadBalancer", {
  loadBalancerArn: "my-alb-arn",
  securityGroupId: "my-alb-sg-id", 
  securityGroupAllowsAllOutbound: false,
});

// create our task definition, container, and add port mappings
const taskDefinition = new ecs.FargateTaskDefinition(this, "TD2");
const container = taskDefinition.addContainer("Service2Container", {
  image: ecs.ContainerImage.fromRegistry("nginx"),
  environment: {"MY_ENVIRONMENT_VAR": "foo"},
  memoryLimitMiB: 512,
  cpu: 256
});
container.addPortMappings({containerPort: 80})

// Create a service -- this instance isn't particularly complex, since we only have
// one container image per service that we need to worry about. 
// We'll reuse the cluster from our first service but use our newly created 
// security group for this purpose. 
const service2 = new ecs.FargateService(this, "service2", {
  cluster: importedCluster,
  taskDefinition
});

// Set up security group ingress/egress between LB and new service
service2.connections.allowFrom(loadBalancer, ec2.Port.tcp(80));
loadBalancer.connections.allowTo(service2, ec2.Port.tcp(80));
// Allow connections between services if necessary. This would require importing
// any other service by attributes and ensuring that you specify its security group when
// importing the cluster. 

// Create a new target group for the service and add it to the listener, specifying
// a path, host, or IP pattern and a priority. 
const tg = new ApplicationTargetGroup(this, "targetgroup", {
  targets: [service2],
  protocol: ApplicationProtocol.HTTP,
  vpc,
});
listener.addTargetGroups("targetgroup", {
  targetGroups: [tg],
  pathPattern: "/app2/*",
  priority: 100
});
```

**Comment by rmmeans** [comment]
@bvtujo - thanks for the detailed reply. 

For your questions, each of our microservices deploy separately. This is a pattern we've been doing for awhile using plain cloudformation and sharing an ALB, cluster, load balancer SG, etc.

I have been working on my own construct for now that allows me to manage it like we have done in the past. The routing is done on host headers at the ALB, so each fargate target group can pick whatever port mappings makes sense for that service and the ALB just dishes out the requests to the proper microservice target group based on the host header of the microservice.

In my construct use case, I am creating the target group, binding the target group to the existing listener on the existing ALB, creating the fargate task, container def,  and service. Finally, I ensure the traffic from the ALB's already existing SG is able to pass traffic to the App's SG.

All-in-all, I have actually found this process very easy to create my own constructs for the way we deploy infrastructure. I'm not sure how much I would really need / use a ecs-patterns construct now that i've seen how easy this is to do myself.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts (hunk 7 above)
Comment (bvtujo) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: done--is this what you had in mind?

**Comment by pkandasamy91** [comment]
> @rmmeans Thanks for the use case details. I'd also be interested in your major deployment concerns and strategies–do you typically need to deploy microservices one at a time, or is it possible to deploy the full breadth of services you require in a single CFN stack? Do you need assurances that deploying a new service won't result in ALB downtime?
> 
> The patterns currently don't support creating multiple services on the same port, as doing so requires that an existing listener and load balancer security group also be imported, but you can hook this up quite straightforwardly with a few lines of code using other CDK modules.
> 
> For an application load balanced Fargate service, the following will do the trick. You will need to import a few things other than the load balancer from the other stack, however, including:
> 
>     * VPC
> 
>     * cluster
> 
>     * load balancer security group
> 
>     * security groups of any other services running in the cluster which your service needs to communicate with
> 
>     * listener
> 
>     * load balancer
> 
> 
> ```js
> // take care of importing all required constructs 
> const vpc = ec2.Vpc.fromLookup(this, "vpc", {vpcId: "my-vpc-id"});
> const lbSecurityGroup = ec2.SecurityGroup.fromSecurityGroupId(this, "sgId", "my-alb-sg-id");
> const importedCluster = ecs.Cluster.fromClusterAttributes(this, "clusterImport", {
>   clusterName: "my-cluster",
>   vpc,
>   securityGroups: [ec2.SecurityGroup.fromSecurityGroupId(this, "sg", "other-service-sg")]; //and other security groups on the cluster
> });
> const listener = elbv2.ApplicationListener.fromApplicationListenerAttributes(this, "Listener", {
>   listenerArn: "my-application-listener-arn",
>   securityGroup: lbSecurityGroup,
> });
> const alb = elbv2.ApplicationLoadBalancer.fromApplicationLoadBalancerAttributes(this, "loadBalancer", {
>   loadBalancerArn: "my-alb-arn",
>   securityGroupId: "my-alb-sg-id", 
>   securityGroupAllowsAllOutbound: false,
> });
> 
> // create our task definition, container, and add port mappings
> const taskDefinition = new ecs.FargateTaskDefinition(this, "TD2");
> const container = taskDefinition.addContainer("Service2Container", {
>   image: ecs.ContainerImage.fromRegistry("nginx"),
>   environment: {"MY_ENVIRONMENT_VAR": "foo"},
>   memoryLimitMiB: 512,
>   cpu: 256
> });
> container.addPortMappings({containerPort: 80})
> 
> // Create a service -- this instance isn't particularly complex, since we only have
> // one container image per service that we need to worry about. 
> // We'll reuse the cluster from our first service but use our newly created 
> // security group for this purpose. 
> const service2 = new ecs.FargateService(this, "service2", {
>   cluster: importedCluster,
>   taskDefinition
> });
> 
> // Set up security group ingress/egress between LB and new service
> service2.connections.allowFrom(loadBalancer, ec2.Port.tcp(80));
> loadBalancer.connections.allowTo(service2, ec2.Port.tcp(80));
> // Allow connections between services if necessary. This would require importing
> // any other service by attributes and ensuring that you specify its security group when
> // importing the cluster. 
> 
> // Create a new target group for the service and add it to the listener, specifying
> // a path, host, or IP pattern and a priority. 
> const tg = new elbv2.ApplicationTargetGroup(this, "targetgroup", {
>   targets: [service2],
>   protocol: ApplicationProtocol.HTTP,
>   vpc,
> });
> listener.addTargetGroups("targetgroup", {
>   targetGroups: [tg],
>   pathPattern: "/app2/*",
>   priority: 100
> });
> ```

@bvtujo This is awesome! Would you mind adding this to the examples repo?

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts [hunk 10]
```diff
@@ -297,12 +300,14 @@ export abstract class ApplicationLoadBalancedServiceBase extends cdk.Construct {
       internetFacing
     };
 
-    this.loadBalancer = props.loadBalancer !== undefined ? props.loadBalancer : new ApplicationLoadBalancer(this, 'LB', lbProps);
+    this.loadBalancer = props.loadBalancer !== undefined ? props.loadBalancer as ApplicationLoadBalancer
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: Wait, this is dangerous. Why is the input argument changed to `IApplicationLoadBalancer`, but it's downcast to a concrete class here?

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts [hunk 11]
```diff
@@ -297,12 +300,14 @@ export abstract class ApplicationLoadBalancedServiceBase extends cdk.Construct {
...
+                        : new ApplicationLoadBalancer(this, 'LB', lbProps);
 
     if (props.certificate !== undefined && props.protocol !== undefined && props.protocol !== ApplicationProtocol.HTTPS) {
       throw new Error('The HTTPS protocol must be used when a certificate is given');
     }
-    const protocol = props.protocol !== undefined ? props.protocol : (props.certificate ? ApplicationProtocol.HTTPS : ApplicationProtocol.HTTP);
+    const protocol = props.protocol !== undefined ? props.protocol :
+    (props.certificate ? ApplicationProtocol.HTTPS : ApplicationProtocol.HTTP);
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: Indentation please.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts [hunk 12]
```diff
@@ -268,7 +270,7 @@ export abstract class NetworkLoadBalancedServiceBase extends cdk.Construct {
       internetFacing
     };
 
-    this.loadBalancer = props.loadBalancer !== undefined ? props.loadBalancer : new NetworkLoadBalancer(this, 'LB', lbProps);
+    this.loadBalancer = props.loadBalancer !== undefined ? props.loadBalancer as NetworkLoadBalancer : new NetworkLoadBalancer(this, 'LB', lbProps);
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/network-load-balanced-service-base.ts]: Same comment.

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-load-balancer.ts [hunk 13]
```diff
@@ -38,6 +38,14 @@ export interface NetworkLoadBalancerAttributes {
...
    */
   readonly loadBalancerDnsName?: string;
+
+  /**
+   * The VPC to associate with the load balancer.
+   *
+   * @default - When not provided, listeners cannot be created on imported load
+   * balancers.
+   */
+  readonly loadBalancerVpc?: ec2.IVpc;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-load-balancer.ts]: On the application load balancer this prop is called `vpc`, why is it called `loadBalancerVpc` here? Please make both the same. `vpc` will suffice.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts (hunk 10 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: To completely explain: this cast is not necessarily correct. For example, for an imported load balancer this cast is DEFINITELY NOT correct, as an `ApplicationLoadBalancer` is going to have methods that an `ImportedApplicationLoadBalancer` does not have, and if you try to use those you will get crashes.

If the problem is that `this.loadBalancer` has been typed as an `ApplicationLoadBalancer` and so you must trick the type system, this is not the way to do it! A way to "properly" handle this case while not breaking backwards compatibility would be something like this:

```ts
class {
  private readonly _applicationLoadBalancer?: ApplicationLoadBalancer;

  constructor() {
    if (props.loadBalancer instanceof ApplicationLoadBalancer) {
      this._applicationLoadBalancer = props.loadBalancer;
    }
  } 

  public get applicationLoadBalancer(): ApplicationLoadBalancer {
    if (!this._applicationLoadBalancer) {
      throw new Error('.applicationLoadBalancer can only accessed if the class was constructed with an owned load balancer, not an imported one');
    } 
    return this._applicationLoadBalancer;
  }
}
```

That is, the type-safe API promises things we don't want to guarantee anymore, so now we have to transform it into a run-time check.

Not ideal, but the best we can do given the contracts we uphold.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts (hunk 11 above)
Comment (bvtujo) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: Resolved in 57cd88a0

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts (hunk 10 above)
Comment (bvtujo) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts]: Thanks for the detailed feedback, this was really helpful in understanding the guards we put against backwards compatibility.

**Pull request merged**